File: src/semi_heap_sort.rb
class SemiHeapSort
    def self.check_tournament(t)
        t.each_with_index(:upper) do |e, row, col|
            raise "#{row}, #{col} must be 1, 0 or nil and is #{e}" unless [0, 1, nil].include? e
            if row == col
                raise "#{row}, #{col} ought to be nil" unless e.nil?
            else
                raise "mismatch in #{row}, #{col} and #{col}, #{row}" if t[col,row] == e
            end
        end
    end

    def self.beats?(a_i, a_j, &comparator)
        return false if a_i.nil?
        return true if a_j.nil?
        result = comparator.call(a_i, a_j)
        raise "invalid comparison result #{result} - must be -1 or 1" if result != 1 and result != -1
        return result == -1
    end

    def self.max(a_i, a_j, a_k, &comparator)
        return a_i if beats?(a_i, a_j, &comparator) && beats?(a_i, a_k, &comparator)
        return a_j if beats?(a_j, a_i, &comparator) && beats?(a_j, a_k, &comparator)
        return a_k if beats?(a_k, a_i, &comparator) && beats?(a_k, a_j, &comparator)
        return nil
    end

    def self.is_max?(a_i, a_j, a_k, &comparator)     # if i beats j and k
        max(a_i, a_j, a_k, &comparator) == a_i
    end

    def self.is_semi_max?(a_i, a_j, a_k, &comparator)    # if i is semi max, if neither j nor k are max
        !is_max?(a_j, a_i, a_k, &comparator) && !is_max?(a_k, a_i, a_j, &comparator)
    end

    def self.left(index)
        2 * index + 1
    end

    def self.right(index)
        2 * index + 2
    end

    def self.replace(a, i, &comparator)
        if a[left(i)].nil? && a[right(i)].nil?
            a[i] = nil
        elsif beats?(a[i], a[left(i)], &comparator) && beats?(a[left(i)], a[right(i)], &comparator)
            a[i] = a[left(i)]
            replace(a, left(i), &comparator)
        else
            a[i] = a[right(i)]
            replace(a, right(i), &comparator)
        end
    end

    def self.semi_heapify(a, i, &comparator)
        if not is_semi_max?(a[i], a[left(i)], a[right(i)], &comparator)
            winner = max(a[i], a[left(i)], a[right(i)], &comparator)
            raise "no winner" if winner.nil?
            winner_index = winner == a[left(i)] ? left(i) : right(i)
            a[i], a[winner_index] = a[winner_index], a[i]
            semi_heapify(a, winner_index, &comparator)
        end
    end

    def self.build_semi_heap(a, &comparator)
        a = a + [nil] * a.size
        
        ((a.size - 1) / 4).floor.downto(0) do |i|
            semi_heapify(a, i, &comparator)
        end
        a
    end

    def self.sort(elements, &comparator)
        raise "no comparison block given" unless block_given?

        a = build_semi_heap(elements, &comparator)

        result = []
        while a[left(0)] != nil || a[right(0)] != nil
            result << a[0]
            replace(a, 0, &comparator)
        end
        result << a[0]
    end
end

File: src/semi_heap_sort.js
var SemiHeapSort = {
    beats: function(a_i, a_j, comparator) {
        if (a_i == null) {
            return false;
        }
        if (a_j == null) {
            return true;
        }
        var result = comparator(a_i, a_j);
        if (result != 1 & result != -1) {
            throw "invalid comparison result " + result + " - must be -1 or 1";
        }
        return result == -1;
    },

    max: function(a_i, a_j, a_k, comparator) {
        if (this.beats(a_i, a_j, comparator) && this.beats(a_i, a_k, comparator)) {
            return a_i;
        }
        if (this.beats(a_j, a_i, comparator) && this.beats(a_j, a_k, comparator)) {
            return a_j;
        }
        if (this.beats(a_k, a_i, comparator) && this.beats(a_k, a_j, comparator)) {
            return a_k;
        }
        return null;
    },

    is_max: function(a_i, a_j, a_k, comparator) {
        return this.max(a_i, a_j, a_k, comparator) == a_i;
    },

    is_semi_max: function(a_i, a_j, a_k, comparator) {
        return !this.is_max(a_j, a_i, a_k, comparator) && !this.is_max(a_k, a_i, a_j, comparator);
    },

    left: function(index) {
        return 2 * index + 1;
    },

    right: function(index) {
        return 2 * index + 2;
    },

    replace: function(a, i, comparator) {
        if (a[this.left(i)] == null && a[this.right(i)] == null) {
            a[i] = null;
        } else if (this.beats(a[i], a[this.left(i)], comparator) && this.beats(a[this.left(i)], a[this.right(i)], comparator)) {
            a[i] = a[this.left(i)];
            this.replace(a, this.left(i), comparator);
        } else {
            a[i] = a[this.right(i)];
            this.replace(a, this.right(i), comparator);
        }
    },

    semi_heapify: function(a, i, comparator) {
        if (!this.is_semi_max(a[i], a[this.left(i)], a[this.right(i)], comparator)) {
            var winner = this.max(a[i], a[this.left(i)], a[this.right(i)], comparator);
            if (winner == null) {
                throw "no winner";
            }
            var winner_index = winner == a[this.left(i)] ? this.left(i) : this.right(i);
            var temp = a[i];
            a[i] = a[winner_index];
            a[winner_index] = temp;
            this.semi_heapify(a, winner_index, comparator);
        }
    },

    build_semi_heap: function(a, comparator) {
        var result = a.slice();
        result = result.concat(new Array(a.length).fill(null));
        for (var i = Math.floor((result.length - 1) / 4); i >= 0; i--) { 
            this.semi_heapify(result, i, comparator);
        }

        return result;
    },

    sort: function(elements, comparator) {
        var a = this.build_semi_heap(elements, comparator);

        var result = [];
        while (a[this.left(0)] != null || a[this.right(0)] != null) {
            result.push(a[0]);
            this.replace(a, 0, comparator);
        }
        result.push(a[0]);
        return result;
    }
}
